src/codec.rs

```diff
@@ -466,17 +486,29 @@ where
 	// However, if `T` doesn't contain any `Bytes` then this extra allocation is
 	// technically unnecessary, and we can avoid it by tracking the position ourselves
 	// and treating the underlying `Bytes` as a fancy `&[u8]`.
-	let mut input = BytesCursor { bytes, position: 0 };
+	let mut input = BytesCursor::new(bytes);
 	T::decode(&mut input)
 }
 
 #[cfg(feature = "bytes")]
 impl Decode for bytes::Bytes {
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
-		input.scale_internal_decode_bytes()
+		let len = <Compact<u32>>::decode(input)?.0 as usize;
+		if input.__private_bytes_cursor().is_some() {
+			input.on_before_alloc_mem(len)?;
+		}
+
+		if let Some(bytes_cursor) = input.__private_bytes_cursor() {
+			bytes_cursor.decode_bytes_with_len(len)
+		} else {
```

Comment:
```suggestion
		if let Some(bytes_cursor) = input.__private_bytes_cursor() {
			input.on_before_alloc_mem(len)?;
			bytes_cursor.decode_bytes_with_len(len)
		} else {
```

is it more readable?

Reply:
ah indeed, my bad